source/common/router/config_impl.cc

```diff
@@ -1439,6 +1445,17 @@ VirtualHostImpl::VirtualHostImpl(
     hedge_policy_ = virtual_host.hedge_policy();
   }
 
+  if (!virtual_host.request_mirror_policies().empty()) {
+    for (const auto& mirror_policy_config : virtual_host.request_mirror_policies()) {
+      shadow_policies_.push_back(std::make_shared<ShadowPolicyImpl>(mirror_policy_config));
+    }
+  }
+
+  // Inherit policies from the route config.
+  if (shadow_policies_.empty()) {
+    shadow_policies_ = global_route_config_.shadowPolicies();
+  }
```

Comment:
Another idea is creating a method called "mostSpecificShadowPolicies()" to the route, it will return the shadow policies from the level which `shaowPolicies().size() != 0`. Then you needn't duplicated same pointer from the top level to the last level. But yea, shared_prt cost is low, maybe totally fine. We can see the maintainer's opinion here.

Reply:
Yeah I would look at how `mostSpecificPerFilterConfig` works which I think is similar to what this does?

Reply:
Thanks for the input!

`RouteAction` `VirtualHost` and `RouteConfiguration` are three layers design.
`RouteAction` has a private member which points to `VirtualHost`
 `VirtualHost` has a private member which points to  `RouteConfiguration` 
i.e., `RouteAction` doesn't know `RouteConfiguration` 

I think it's a good encapsulation at this moment. 
If We want to do `mostSpecificPerFilterConfig` that means we need to let `RouteAction` be able to "know" `RouteConfiguration`
That's doable but you might know better than me to determine if this is a good time to connect the bridge.

As for complexity, the config is one time setup, and usually not many mirror policies compared to the number of traffic. I *think* it's okay.

What do you think @soulxu @snowp ?

Reply:
Or you can just change the RouteAction's shadowPolices() (maybe a new method called mostSpecificShadowPolicies()) method as:
```
return shadow_policies_.size() == 0 ? virtualHost().shadowPolicies() : shadow_policies_;
```

Then you have VirtualHost's shadowPolicies return as:
```
return shadow_policies_.size() == 0 ? routeConfiguration().shadowPolicies() : shadow_policies_;
```

not sure if this is make sense or not.

Reply:
That's a good idea for implementation. It doesn't break the abstraction we want to keep, which is good. Thanks, soulxu

However I'm still hesitated to do that given one might be misused the member `shadow_policies_` which doesn't not reflect the real policies. On the other hand, it doesn't align what we did with other fields like `retry_policy` and `hedge_policy`. Therefore when we manipulates those fields in multiple layers, we need to pay more attention on the handling. 

Does it make sense @soulxu @snowp ?

Reply:
> However I'm still hesitated to do that given one might be misused the member `shadow_policies_` which doesn't not reflect the real policies. On the other hand, it doesn't align what we did with other fields like `retry_policy` and `hedge_policy`. Therefore when we manipulates those fields in multiple layers, we need to pay more attention on the handling.
> 

Those make sense to me also. I don't have strong feel on this, both works for me. Thanks for taking time to look at my comment also!

Reply:
@snowp Does this thinking process make sense to you? If yes, could you please help to move this forward? Thanks.